core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -835,10 +719,120 @@ int Cuda::impl_is_initialized() {
 }
 
 void Cuda::impl_initialize(InitializationSettings const &settings) {
+  auto &cuda_singleton     = Impl::CudaInternal::singleton();
```

Comment:
Move it closer to its use.  We should avoid constructing instances (singleton comprised) before everything else is ready.